File: architecture-skills/object-oriented-programming-paradigm/examples/dependency-inversion/dependency-inversion.py
"""Dependency Inversion Principle in Python.

OrderService owns the Notifier protocol and receives an implementation
through its constructor — never `new SmtpNotifier(...)`. main() is the
composition root that wires ConsoleNotifier or RecordingNotifier.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class Order:
    id: str
    customer_email: str


class Notifier(Protocol):
    """Abstraction owned by policy: shaped to what place() needs."""

    def notify(self, email: str, message: str) -> None: ...


class OrderService:
    def __init__(self, notifier: Notifier) -> None:
        self._notifier = notifier

    def place(self, order: Order) -> None:
        print(f"placed {order.id} for {order.customer_email}")
        self._notifier.notify(
            order.customer_email,
            f"order {order.id} confirmed",
        )


# --- Low-level details: depend on the policy's protocol, not vice versa ---


class ConsoleNotifier:
    def notify(self, email: str, message: str) -> None:
        print(f"notify {email}: {message}")


class RecordingNotifier:
    """Test double: captures notifications without I/O."""

    def __init__(self) -> None:
        self.sent: list[tuple[str, str]] = []

    def notify(self, email: str, message: str) -> None:
        self.sent.append((email, message))


if __name__ == "__main__":
    order = Order(id="ORD-001", customer_email="alice@example.com")

    print("-- ConsoleNotifier at composition root --")
    OrderService(ConsoleNotifier()).place(order)

    print("\n-- RecordingNotifier: policy testable without I/O --")
    recorder = RecordingNotifier()
    OrderService(recorder).place(Order(id="ORD-002", customer_email="bob@example.com"))
    print(f"{len(recorder.sent)} notification(s) recorded")

File: architecture-skills/object-oriented-programming-paradigm/examples/dependency-inversion/dependency-inversion.swift
// Dependency Inversion Principle in Swift.
//
// OrderService owns the Notifier protocol and receives an implementation
// through its initializer — never SmtpNotifier(). main is the composition
// root that wires ConsoleNotifier or RecordingNotifier.

import Foundation

struct Order {
    let id: String
    let customerEmail: String
}

/// Abstraction owned by policy: shaped to what place() needs.
protocol Notifier {
    func notify(email: String, message: String)
}

final class OrderService {
    private let notifier: Notifier

    init(notifier: Notifier) {
        self.notifier = notifier
    }

    func place(_ order: Order) {
        print("placed \(order.id) for \(order.customerEmail)")
        notifier.notify(
            email: order.customerEmail,
            message: "order \(order.id) confirmed",
        )
    }
}

// --- Low-level details: depend on the policy's protocol ---

final class ConsoleNotifier: Notifier {
    func notify(email: String, message: String) {
        print("notify \(email): \(message)")
    }
}

final class RecordingNotifier: Notifier {
    private(set) var sent: [(String, String)] = []

    func notify(email: String, message: String) {
        sent.append((email, message))
    }
}

print("-- ConsoleNotifier at composition root --")
OrderService(notifier: ConsoleNotifier()).place(
    Order(id: "ORD-001", customerEmail: "alice@example.com"),
)

print("\n-- RecordingNotifier: policy testable without I/O --")
let recorder = RecordingNotifier()
OrderService(notifier: recorder).place(
    Order(id: "ORD-002", customerEmail: "bob@example.com"),
)
print("\(recorder.sent.count) notification(s) recorded")
